guidance/library/_gen.py

```diff
@@ -131,58 +126,58 @@ def gen(
-                raise ValueError(f"Could not infer unambiguous tool call prefix for tool {tool.name}")
-            options.append(
-                capture(
-                    Gen(body_regex=regex, stop_regex=quote_regex(tool_call_prefix), max_tokens=max_tokens),
-                    name=f"tool{i}"
+        @guidance(stateless=False, dedent=False)
+        def tool_gen(lm):
+            options = [
+                grammar_gen(regex=regex, stop_regex=gen_stop, save_stop_text=save_stop_text, temperature=temperature, max_tokens=max_tokens)
+            ]
+            for i, tool in enumerate(tools):
+                # Infer a regex that will match the start of a tool call
+                tool_call_prefix = tool.call_grammar.forced_prefix()
+                if len(tool_call_prefix) < 4:
+                    # TODO: alternatively check that the prefix contains the name (case insensitive) of the tool?
+                    # anything shorter is probably far too ambiguous
+                    raise ValueError(f"Could not infer unambiguous tool call prefix for tool {tool.name}")
+                options.append(
+                    grammar_gen(regex=regex, stop_regex=quote_regex(tool_call_prefix), temperature=temperature, max_tokens=max_tokens, name=f"tool{i}"),
                 )
-            )
-        grm = with_temperature(select(options), temperature)
-        initial_token_count = lm.token_count
-        with block(tagged_name):
-            while lm.token_count <= max_tokens + initial_token_count:
-                lm += grm
-                tool_called = False
-                for i in range(len(tools)):
-                    tool_i = f"tool{i}"
-                    if tool_i in lm:
-                        tool_called = True
-                        if hide_tool_call:
-                            temp_lm = lm + tools[i].call_grammar
-                            with block("tool_call"):
-                                temp_lm += tools[i].tool_call()
-                            lm += temp_lm["tool_call"]
-                        else:
-                            lm += tools[i].call_grammar + tools[i].tool_call()
-                        lm = lm.remove(tool_i)
-                if not tool_called:
-                    lm += suffix
-                    break
-        return lm
-
-    pattern = Gen(body_regex=regex, stop_regex=gen_stop, save_stop_text=save_stop_text, capture_name=tagged_name, max_tokens=max_tokens)
-
-    # define any capture group for non-tool calls
-    if name is not None and tools is None:
-        pattern = capture(pattern, name=tagged_name)
-    lm += with_temperature(pattern + suffix, temperature)
-
-    logger.debug(f"finish gen")
-    return lm
+
+            grm = with_temperature(select(options), temperature)
+            initial_token_count = lm.token_count
+            tagged_name = "__LIST_APPEND:" + name if list_append and name is not None else name
```

Comment:
we seem to be already adding `__LIST_APPEND:` when serializing; is there a problem? (I'm probably missing something)